Show proper error message even if fetching origin user id fails

NobodysNightmare opened the pull request:
I noticed an additional case where we do not show the intended error message for 61612. Reproduction is comparatively easy:

1. Have a working setup using token **exchange**
2. Misconfigure the storage to use the IDP token instead of exchanging for a token with the right audience (storage will respond with 401 on every request)

This way, when opening the files tab, we'd previously not have shown the files, but would have displayed a red error message about an HTTP 500. The error occured while fetching the remote identity, but it would even have happened if we already knew the remote identity for the user.

This PR fixes two issues:
1. We only fetch the remote identity information, if there is no remote identity yet (causing fewer requests to storage in general, but also making reproduction harder: you need to have **no** remote identity)
2. Fetch errors of the remote identity are now only logged, but do not raise an error across the `OpenProject::Notification` boundary. This allows the UI to show an appropriate error message.

# Ticket
https://community.openproject.org/projects/cross-application-user-integration-stream/work_packages/61986

NobodysNightmare committed:
Ensure that we only query origin_user_id when it's missing

Previously we queried the remote storage whenever we exchanged
a token for a user, just to confirm that we still know their
origin_user_id.

Now we will only query the origin_user_id when the remote identity
is new. If we already have a remote identity, we will not try to
update it and will not make additional web requests.

NobodysNightmare committed:
Failed HTTP requests do not raise errors from TOKEN_OBTAINED_EVENT

We will now propagate HTTP errors upwards as failed service results,
top-most they are only logged, but will not raise an error anymore.

We can do this, because a missing remote identity will be shown to
the user as an error anyways. On the other hand raising an error
from the OpenProject::Notifications breaks the code performing
the original token fetching and prevents display of the nice error
message in the UI.

ba1ash commented on app/services/remote_identities/create_service.rb:
```diff
@@ -45,8 +45,11 @@ def initialize(user:, integration:, token:)
     def call
-      @model.origin_user_id = @integration.extract_origin_user_id(@token)
-      if @model.changed?
+      if @model.new_record?
```
:red_circle: 
I am not sure about that change. At least the following case is possible which I consider incorrect:
1. OP user has OAuth connection to NC(OAuthClientToken and RemoteIdentity)
2. Then OP user removes OAuthClientToken through UI, RemoteIdentity remains.
3. User goes though OAuth grant flow, but logs in as a different user on NC side.
4. Exisiting RemoteIdentity is not updated and contains wrong origin_user_id.

NobodysNightmare replied:
Agreed. That case is a problem... Let me think about it.

NobodysNightmare replied:
Here's my idea, tell me what you think:

* `RemoteIdentities::CreateService` accepts a flag that forces an update of the remote identity
* ConnectionManager sets that flag, SSO doesn't

In general I think all of this is just a workaround. A really nice thing would be to offer the user management/transparency about remote identities, the same as we show for access tokens for other applications. But as long as we don't have that, this approach could solve most of the issues we see:

* the case you described does not occur anymore
* we do not flood the storage with queries about the user identity from SSO; the OAuth 2.0 case makes sure that we don't do that anyways, by not calling the `CreateService` every time

ba1ash replied:
I'm fine with this idea.

NobodysNightmare replied:
implemented it

NobodysNightmare committed:
Make sure that remote identity is updated from ConnectionManager

The main intention of switching to the approach of requiring the record
remote identity to be new/unpersisted, was to avoid unnecessary calls to
extract_origin_user_id, which results in a web call. However, the ConnectionManager
already ensures to only call the RemoteIdentities::CreateService is a small number
of cases, so it should be fine to always update when calling it.

We also identified one corner case, where it might be preferable to do that
update from the ConnectionManager: Users could log-in to a storage with
a different account after removing a previous OAuth connection. Without the
update, we'd not change the origin_user_id properly. The risk of that happening
is much lower in an SSO scenario.

ba1ash reviewed (approved):
LGTM.

ba1ash commented on app/services/remote_identities/create_service.rb:
```diff
@@ -32,21 +32,25 @@ module RemoteIdentities
+        user_id = @integration.extract_origin_user_id(@token)
+        return user_id if user_id.failure?
+
+        @model.origin_user_id = user_id.result
```
Would it make sense to check if model changed to avoid unnecessary event emission? 
```suggestion
        @model.origin_user_id = user_id.result
        return ServiceResult.success(result: @model) unless @model.changed?
```

NobodysNightmare replied:
done

ba1ash replied:
Thanks!
